BPOR/BPOR.Rms/Controllers/VolunteerController.cs

```diff
@@ -11,6 +12,123 @@ namespace BPOR.Rms.Controllers;
 
 public class VolunteerController(ParticipantDbContext context) : Controller
 {
+    public IActionResult Consent()
+    {
+        return View(new VolunteerContactConsentViewModel());
+    }
+
+    [HttpPost]
+    public IActionResult Consent(VolunteerContactConsentViewModel model)
+    {
+        if (!model.AgreedToContactConsent)
+        {
+            ModelState.AddModelError("AgreedToContactConsent", "Confirm that the Privacy and Data Sharing Policy has been read and understood before giving consent");
+        }
+
+        if (ModelState.IsValid)
+        {
+            return RedirectToAction("Create");
+        }
+
+        return View(model);
+    }
+
+    public IActionResult Create()
+    {
+        var model = new VolunteerFormViewModel();
+        return View(model);
+    }
+
+    // POST: Study/Create
+    // To protect from overposting attacks, enable the specific properties you want to bind to.
+    // For more details, see http://go.microsoft.com/fwlink/?LinkId=317598.
+    [HttpPost]
+    // [ValidateAntiForgeryToken]
+    public async Task<IActionResult> Create(
+        [Bind("FirstName,LastName,DateOfBirth,PostCode,AddressLine1,AddressLine2,AddressLine3,AddressLine4,Town,PreferredContactMethod,Email,LandLine,Mobile" +
+        ",SexRegisteredAtBirth,GenderIdentitySameAsBirth,EthnicGroup,EthnicBackground,LongTermConditionOrIllness,AreasOfResearch")]
+        VolunteerFormViewModel model, string action)
+    {
+
+        if (String.IsNullOrEmpty(model.LandLine) && String.IsNullOrEmpty(model.Mobile))
+        {
+            ModelState.AddModelError("LandLine", "At least one of either a Landline or Mobile number must be provided");
+        }
+        if (model.PreferredContactMethod == "Email" && String.IsNullOrEmpty(model.Email))
+        {
+            ModelState.AddModelError("Email", "Email must be provided when preferred contact method is email");
+        }
+
+        ValidateDateOfBirth(model.DateOfBirth);
+
+        if (!String.IsNullOrEmpty(model.Email))
+        {
+            DoesUserEmailExistInDatabase(model.Email);
+        }
+
+        if (model.DateOfBirth.HasValue && model.PostCode.HasValue && !String.IsNullOrEmpty(model.LastName))
+        {
+            DoesPostcodeSurnameDoBComboExist(model.PostCode.Value.ToString(), model.LastName, model.DateOfBirth);
+        }
+
+        if (ModelState.IsValid)
+        {
+            return RedirectToAction(nameof(VolunteerController.AccountSuccess));
+        }
+
+        return View(model);
+    }
+
+    private void DoesPostcodeSurnameDoBComboExist(string postCode, string lastName, GovUkDate dateOfBirth)
```

Comment:
this method can be marked as async and return a Task
then use await (line 86).... FirstOrDefaultAsync() (line 90)

Reply:
Thanks @petecorriepa I've updated these now to use async tasks